Contrib/FreeWilson/freewilson.py

```diff
@@ -0,0 +1,486 @@
+            rejected_pred += 1
+            continue
+        good_pred+=1
+        smiles = set([g.smiles for g in groups]) # remove dupes
+        smi = ".".join(set([g.smiles for g in groups]))
+        try:
+            mol = molzip(Chem.MolFromSmiles(smi))
+        except:
+            rejected_bad += 1
+            continue
+            
+        rejected = False
+        if mol_filter and not mol_filter(mol):
+            rejected_filters += 1
+            continue
+
+        out_smi = Chem.MolToSmiles(mol)
+        yield FreeWilsonPrediction(pred, out_smi, groups)
+        wrote += 1
+    logging.info(f"Wrote {wrote} results out of {num_products}\n\tIn Training set: {in_training_set}\n\tBad MW: {rejected_mw}\n\tBad Pred: {rejected_pred}\n\tBad Filters: {rejected_filters}\n\tBad smi: {rejected_bad}\n\tmin mw: {min_mw}\n\tmax mw: {max_mw}\n\t\n\tmin pred: {min_pred}\n\tmax pred: {max_pred}")
+
+    
+def FWBuild(fw: FreeWilsonDecomposition,
+            pred_filter=None,
+            mw_filter=None,
+            mol_filter=None) -> Generator[FreeWilsonPrediction,None,None]:
+    """Enumerate the freewilson decomposition and return their predictions
+
+       :param fw: FreeWilsonDecomposition generated from FWDecompose
+       :param pred_filter: return True if the prediction is in a desireable range
+                           e.g.  lambda pic50: pic50>=8
+       :param mw_filter: return True if the enumerated molecular weight is in a desireable rrange
+                           e.g. lambda mw: 150 < mw < 550
+       :param mol_filter: return True if the molecule is ok to be enumerated
+                           e.g. lambda mol: -3 < Descriptors.MolLogp(mol) < 5
+    """
+    blocker = rdBase.BlockLogs()
+    # check groups for cycles
+    cycles = set()
+    rgroups_no_cycles = defaultdict(list)
+    rgroup_cycles = defaultdict(list)
+    for key, rgroup in fw.rgroups.items():
+        if key == 'Core':
+            rgroups_no_cycles[key] = rgroup
+            continue
+        no_cycles = rgroups_no_cycles[key]
+        for g in rgroup:
+            if len(g.dummies) > 1:
+                cycles.add(g.dummies)
+                rgroup_cycles[g.dummies].append(g)
+            else:
+                no_cycles.append(g)
+
+    logging.info("Enumerating rgroups with no broken cycles...")
+    for k,v in rgroups_no_cycles.items():
+        logging.info(f"\t{k}\t{len(v)}")
+    # do the ones that have no cycles first
+    rgroups = [rgroup for key, rgroup in sorted(rgroups_no_cycles.items())]
+    # core is always first
+    for res in _enumerate(rgroups,
+                          fw,
+                          pred_filter=pred_filter,
+                          mw_filter=mw_filter,
+                          mol_filter=mol_filter):
+        yield res
+
+    # iterate on rgroups with cycles
+    #  basically only let one set of RGroups show up once.
+    indices = set()
+    for k in fw.rgroups.keys():
```

Comment:
```suggestion
    for k in fw.rgroups:
```

Reply:
It's been a long time since I've had to use .keys() to iterate through a dictionary...